be/src/exec/workgroup/work_group.h

```diff
@@ -10,82 +10,132 @@
-    WG_DEFAULT = 1,  // default work group
-    WG_REALTIME = 2, // realtime work group, maybe reserved beforehand
-};
+using WorkGroupType = TWorkGroupType::type;
 // WorkGroup is the unit of resource isolation, it has {CPU, Memory, Concurrency} quotas which limit the
 // resource usage of the queries belonging to the WorkGroup. Each user has be bound to a WorkGroup, when
 // the user issues a query, then the corresponding WorkGroup is chosen to manage the query.
 class WorkGroup {
 public:
-    WorkGroup(const std::string& name, int id, size_t cpu_limit, size_t memory_limit, size_t concurrency,
-              WorkGroupType type);
+    WorkGroup(const std::string& name, int64_t id, int64_t version, size_t cpu_limit, double memory_limit,
+              size_t concurrency, WorkGroupType type);
+    WorkGroup(const TWorkGroup& twg);
     ~WorkGroup() = default;
 
+    TWorkGroup to_thrift() const;
+    TWorkGroup to_thrift_verbose() const;
     void init();
 
-    starrocks::MemTracker* mem_tracker() { return _mem_tracker.get(); }
-    starrocks::pipeline::DriverQueue* driver_queue() { return _driver_queue.get(); }
+    MemTracker* mem_tracker() { return _mem_tracker.get(); }
+    pipeline::DriverQueue* driver_queue() { return _driver_queue.get(); }
+
+    int64_t id() const { return _id; }
+
+    int64_t version() const { return _version; }
+
+    const std::string& name() const { return _name; }
 
-    int id() const { return _id; }
-    int get_cpu_priority() {
-        // TODO: implement cpu priority computation
-        return 0;
+    static constexpr int64 DEFAULT_WG_ID = 0;
+    static constexpr int64 DEFAULT_VERSION = 0;
+    bool try_offer_io_task(const PriorityThreadPool::Task& task);
+
+public:
+    // increase num_driver when the driver is attached to the workgroup
+    void increase_num_drivers() {
+        ++_num_drivers;
+        ++_acc_num_drivers;
     }
-    int get_io_priority() {
-        // TODO: implement io priority computation
-        return 0;
+    // decrease num_driver when the driver is detached from the workgroup
+    void decrease_num_drivers() { --_num_drivers; }
+
+    // mark the workgroup is deleted, but at the present, it can not be removed from WorkGroupManager, because
+    // 1. there exists pending drivers
+    // 2. there is a race condition that a driver is attached to the workgroup after it is marked del.
+    void mark_del() {
+        bool expect_false = false;
+        if (_is_marked_del.compare_exchange_strong(expect_false, true)) {
+            static constexpr seconds expire_seconds{120};
+            _vacuum_ttl = duration_cast<milliseconds>(steady_clock::now().time_since_epoch() + expire_seconds).count();
+        }
+    }
+    // no drivers shall be added to this workgroup
+    bool is_marked_del() const { return _is_marked_del.load(std::memory_order_acquire); }
+
+    // a workgroup should wait several seconds to be cleaned safely.
```

Comment:
Is it simillar to the concept of TTL in TCP?

Reply:
different, alter a workgroup is altered, the old version should be uninstalled, while the new version should be installed, the old version should wait for some while after there exists no pending driver joins it.